common/achievement/template.ts

```diff
@@ -88,4 +93,165 @@ async function getTemplatesByMetrics(metricsForAction: Metric[]) {
+
+        // TODO: How to evaluate the aggregators?
+    }
+
+    // TODO: Does the template.type imply any prerequisities?
+
+    const sampleResult = await swan.parse(template.condition)(Object.fromEntries(aggregations.map((it) => [it, 0])));
+    if (typeof sampleResult !== 'boolean') {
+        throw new PrerequisiteError(
+            `AchievementTemplate condition does not evaluate to a boolean - This could be as it references to non existent aggregations`
+        );
+    }
+
+    // ---- Template Group --------
+    // We generally assume that if one template is activated that the whole group is supposed to be activated
+    // Thus also inconsistencies of not yet enabled achievement templates might show up here (this is easier to check)
+
+    // Check that groupOrders are sequential without gaps and that they are activated in sequence
+    let currentOrder = 1;
+    for (const groupTemplate of group) {
+        if (groupTemplate.groupOrder !== currentOrder) {
+            throw new PrerequisiteError(`Inconsistency in groupOrder, must be sequential`);
+        }
+
+        if (currentOrder < template.groupOrder && !groupTemplate.isActive) {
+            throw new PrerequisiteError(`Templates of a sequence must be activated in order`);
+        }
+
+        currentOrder += 1;
+    }
+
+    // Check that stepNames are set for a group with multiple steps
+    if (group.length > 1) {
+        for (const groupTemplate of group) {
+            if (!groupTemplate.stepName) {
+                throw new PrerequisiteError(`For templates of a sequence, every group template must have a stepName`);
+            }
+        }
+    }
+
+    // Everything is awesome!
+}
+
+export async function activateAchievementTemplate(id: number) {
+    const template = await getTemplate(id);
+
+    if (template.isActive) {
+        throw new RedundantError('Template is already active');
+    }
+
+    await checkTemplateConsistencyBeforeActivating(template);
+
+    await prisma.achievement_template.update({
+        where: { id },
+        data: { isActive: true },
+    });
+
+    purgeAchievementTemplateCache();
+    logger.info(`Activated AchievementTemplate(${id})`);
+}
+
+export async function deactivateAchievementTemplate(id: number) {
+    const template = await getTemplate(id);
+    const group = await getTemplateGroup(template.group);
+
+    if (!template.isActive) {
+        throw new RedundantError('Template is already inactive');
+    }
+
+    const hasAchivements = (await prisma.user_achievement.count({ where: { templateId: id } })) > 0;
+    if (hasAchivements) {
+        throw new PrerequisiteError(`Cannot deactivate AchievementTemplate as it is already in use`);
+    }
+
+    if (group.length > 1 && group.some((other) => other.groupOrder > template.groupOrder && template.isActive)) {
+        throw new PrerequisiteError(`Cannot deactivate AchievementTemplate with groupOrder ${template.groupOrder} as a following template is still active`);
+    }
```

Comment:
I have no clear opinion if this should be allowed or not.

Reply:
I removed this again, I guess as we only require total order of the group order one could do this